pylint/checkers/typecheck.py

```diff
@@ -1030,7 +1031,21 @@ def visit_attribute(self, node: nodes.Attribute) -> None:
                 if not [
                     n
                     for n in owner.getattr(node.attrname)
-                    if not isinstance(n.statement(future=True), nodes.AugAssign)
+                    if (
+                        # Skip augmented assignments
+                        not isinstance(n.statement(future=True), nodes.AugAssign)
+                        # Skip self-referencing assignments
+                        and n.parent is not node.parent
+                        # Skip assignments that are only type annotations
+                        and not (
+                            isinstance(n.parent, nodes.AnnAssign)
+                            and n.parent.value is None
+                        )
+                        # But allow dataclasses
+                        or brain_dataclasses.is_decorated_with_dataclass(
+                            n.parent.parent
```

Comment:
I can remove the comments if they seem self-evident. We're nested inside a negated list comprehension, so I erred on the side of more help.

Reply:
Do you think it makes sense to create a for loop here? We're calling `n.parent` quite a lot and the readability of this `if` within the comprehension is becoming questionable 😄 

I like the comments personally, makes it easier to understand what's going on.